main_test.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"fmt"
+	"github.com/samber/lo"
+	"github.com/tsaikd/gogstash/internal/randomstring"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/shirou/gopsutil/v3/process"
+)
+
+const ExecutableNamePrefix = "gogstash_test"
+const waitingTime = 100 * time.Millisecond
+
+
+func TestProgramTermination(t *testing.T) {
+	// Start the main program in a separate process
+  executablePath := fmt.Sprintf("./%s_%s", ExecutableNamePrefix, randomstring.RandString(6))
+	_, _ = exec.Command("go", "build", "-o", executablePath).CombinedOutput()
+	defer func() {
+		os.Remove(executablePath)
+	}()
+
+	cmd := exec.Command(executablePath, "--config", "./testdata/config.yaml")
+	if err := cmd.Start(); err != nil {
+		t.Fatalf("Failed to start the program: %v", err)
+	}
+
+	// Get the process ID of the main process
+	mainPID := int32(cmd.Process.Pid)
+
+	// Find all child processes of the main process
+	time.Sleep(waitingTime) // Wait for the program to spawn child processes
+	childProcesses, err := getChildProcesses(mainPID)
+	t.Logf("Children are %s", strings.Join(lo.Map(childProcesses, func(v int32, _ int) string { return strconv.Itoa(int(v)) }), ", "))
+	if err != nil {
+		t.Fatalf("Failed to get child processes: %v", err)
+	}
+	if len(childProcesses) < 1 {
+		t.Fatal("Cannot test if there are no child processes")
+	}
+
+	// Send a SIGINT signal (Ctrl+C) to the main process
+	if err := cmd.Process.Signal(os.Interrupt); err != nil {
+		t.Fatalf("Failed to send SIGINT signal: %v", err)
+	}
+
+	// Wait for a reasonable time for the processes to terminate
+	time.Sleep(waitingTime)
+
+	// Recalling `Processes` to get a fresh information from the OS. Important to force `process.NewProcess` to return valid information.
+	_, err = process.Processes()
+	if err != nil {
+		t.Fatalf("Failed to update processes information: %v", err)
+	}
+
+	// Check whether all child processes have terminated. 
+	for _, pid := range childProcesses {
+		cp, err := process.NewProcess(pid)
+		if err == nil {
+			err = cp.Kill()
+			if err != nil {
+				t.Errorf("Cannot terminate process with PID %d", pid)
+			}
+			t.Errorf("Process with PID %d did not terminate.", pid)
+		}
+	}
+}
+
+func getChildProcesses(parentPID int32) ([]int32, error) {
+	var childPIDs []int32
+	allProcesses, err := process.Processes()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get all processes: %v", err)
+	}
+	for _, p := range allProcesses {
+		ppid, _ := p.Ppid()
```

Comment:
Ignore error?

Reply:
If I could not the ppid of a process it is just meant that this is not the process I am looking for :)

Reply:
It's better to have the explicit error checking, and add some comments if it's reasonable. e.g.
```go
for _, p := range allProcesses {
  ppid, err := p.Ppid()
  if err != nil {
    // blahblah...
    continue
  }
}
```

Reply:
any update?

Reply:
fixed